pennylane/qnode.py

```diff
@@ -605,19 +605,20 @@ def __call__(self, *args, **kwargs):
             # construct the tape
             self.construct(args, kwargs)
 
-        # Execute the tape.
         # If the observable contains a Hamiltonian and the device does not
-        # support Hamiltonians, split tape into multiple tapes and execute
-        # sequentially. In future, this logic should be moved to the device
+        # support Hamiltonians, or if the simulation uses finite shots,
+        # split tape into multiple tapes of diagonalizable known observables.
+        # In future, this logic should be moved to the device
         # to allow for more efficient batch execution.
-        supports_hamiltonian = self.device.supports_observable("Hamiltonian")
+        does_not_support_hamiltonian = not self.device.supports_observable("Hamiltonian")
+        finite_shots = self.device.shots is not None
         hamiltonian_in_obs = "Hamiltonian" in [obs.name for obs in self.qtape.observables]
-        if hamiltonian_in_obs and not supports_hamiltonian:
+        if hamiltonian_in_obs and (does_not_support_hamiltonian or finite_shots):
             try:
                 tapes, fn = qml.transforms.hamiltonian_expand(self.qtape, group=False)
             except ValueError as e:
                 raise ValueError(
-                    "Only a single expectation of a Hamiltonian observable can be returned"
+                    "Only a (single) expectation of a Hamiltonian observable can be returned "

```

Comment:
Reads better, since the error may be raised if one does not use expectation (in which case one wonders why it is complaining about not having a single measurement).

Reply:
I don't really parse the difference, should it read something like "Only a single _measurement_ of a Hamiltonian..."?

Reply:
This is a catch-all error statement: if you have multiple expectations or single non-expectations...Is there a better way to phrase it?

Reply:
I changed it to "single expectation measurements" hoping this makes any difference. It's not so crucial though.

Reply:
Oh okay I see! One could rephrase as something like "The returned measurement of a Hamiltonian observable is  limited to a single expval." or something like this. But fine as-is, I understand now :+1: